src/provider/mod.rs

```diff
@@ -241,6 +257,184 @@ where
+                )?;
+
+        let l2_tx_hash = l1_to_l2_tx_log.inner.txHash;
+
+        let polling_interval_in_seconds =
+            polling_interval_in_seconds.unwrap_or(Duration::from_secs(2));
+        let mut timer = interval(polling_interval_in_seconds);
+        let start = Instant::now();
+
+        loop {
+            timer.tick().await;
+
+            if let Some(timeout) = timeout_in_seconds {
+                if start.elapsed() >= timeout {
+                    return Err(Error::Custom(
+                        "Error waiting for corresponding L2 tx to be included into the block.",
+                    ));
+                }
+            }
+
+            // Wait for the corresponding L2 transaction to be included into the block.
+            let l2_tx_receipt = self.get_transaction_receipt(l2_tx_hash).await.unwrap();
+
+            if let Some(l2_tx_receipt) = l2_tx_receipt {
+                return Ok(l2_tx_receipt);
+            }
+        }
+    }
+}
+
+/// Numerator used in scaling the gas limit to ensure acceptance of `L1->L2` transactions.
+/// This constant is part of a coefficient calculation to adjust the gas limit to account for variations
+/// in the SDK estimation, ensuring the transaction will be accepted.
+const L1_FEE_ESTIMATION_COEF_NUMERATOR: u64 = 12;
+
+/// Denominator used in scaling the gas limit to ensure acceptance of `L1->L2` transactions.
+/// This constant is part of a coefficient calculation to adjust the gas limit to account for variations
+/// in the SDK estimation, ensuring the transaction will be accepted.
+const L1_FEE_ESTIMATION_COEF_DENOMINATOR: u64 = 10;
+
+/// Trait for ZKsync provider with populated wallet
+/// Contains provider methods that need a wallet
+#[cfg_attr(target_arch = "wasm32", async_trait::async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
+pub trait ZksyncProviderWithWallet<T = BoxTransport>:
+    ZksyncProvider<T> + WalletProvider<Zksync>
+where
+    T: Transport + Clone,
+{
+    /// Scales the gas limit to ensure the transaction will be accepted.
+    fn scale_l1_gas_limit(l1_gas_limit: u64) -> u64 {
+        l1_gas_limit * L1_FEE_ESTIMATION_COEF_NUMERATOR / L1_FEE_ESTIMATION_COEF_DENOMINATOR
+    }
+
+    /// Deposits specified L1 token to the L2 address.
+    ///
+    /// ## Parameters
+    ///
+    /// - `token_address`: L1 token address to deposit.
+    /// - `receiver`: receiver L2 address.
+    /// - `amount`: amount to deposit in Wei.
+    /// - `l1_provider`: reference to the L1 provider.
+    ///
+    /// ## Returns
+    ///
+    /// L1 transaction receipt response.
+    /// Hint: use zksync_provider.wait_for_l1_tx method to wait for the corresponding L2 tx to be included into an L2 block.
+    async fn deposit<Tr, N, P>(
+        &self,
+        token_address: Address,
+        receiver: Address,
+        amount: U256,
+        l1_provider: &P,
+    ) -> Result<N::ReceiptResponse, Box<dyn std::error::Error>>
```

Comment:
Added `L1CommunicationError` enum. I decided to use `Custom` enum option with the string explanation instead of a separate option for each error as I don't think that most of the errors in the deposit method will be handled differently.